### DISCUSSION TITLE
Publish to Maven central repository with Github Actions

### DISCUSSION THREAD

**Discussion by arnaudroques** [plantuml/plantuml#769] [General]
I've created [two secrets](https://github.com/plantuml/plantuml/settings/secrets/actions) called `OSSRH_USERNAME` and `OSSRH_PASSWORD` with the appropriate values.
The idea is to update the [pom.xml](https://github.com/plantuml/plantuml/blob/master/pom.xml) file so that the following section:

```
    <scm>
        <connection>scm:svn:svn://svn.code.sf.net/p/plantuml/code/trunk</connection>
        <developerConnection>scm:svn:https://svn.code.sf.net/p/plantuml/code/trunk</developerConnection>
        <url>svn://svn.code.sf.net/p/plantuml/code/trunk</url>
    </scm>
```
uses now this github repository instead of the sourceforge repository.

We could probably create a new action [to publish the packages to maven](https://docs.github.com/en/actions/publishing-packages/publishing-java-packages-with-maven).

We probably have to add this:

```
env:
    OSSRH_USERNAME: ${{ secrets.OSSRH_USERNAME }}
    OSSRH_PASSWORD: ${{ secrets.OSSRH_PASSWORD }}

<server>
  <id>${server}</id>
  <username>xxxxxx</username>
  <password>yyyyyy</password>
</server>
```

somewhere.

For the record, right now, I am running the following commands:

```
mvn clean
mvn deploy
mvn release:clean
mvn release:prepare 
mvn release:perform
```

So the idea is really to automatically publish to Maven central repository when creating a new tag v1.20XX.YY.

@matthew16550 What do you think about it ? I definitively need help here :-)

**Comment by matthew16550** [comment]
Agree it is a good idea.  I can try the publishing next week.

Is there a GPG / PGP signing key?  Might be good to make a new one just for GH Actions.

Suspect we could update the `<scm>` values today regardless of the publish step.  The [requirements](https://central.sonatype.org/publish/requirements/#scm-information) show sample values for GH.

**Reply by arnaudroques** [reply]
> Is there a GPG / PGP signing key? Might be good to make a new one just for GH Actions.

There is a PGP signing key but we should probably make a new one for github.
I guess I have [to follow this](https://docs.github.com/en/authentication/managing-commit-signature-verification/generating-a-new-gpg-key). Is this the only way ?

**Reply by matthew16550** [reply]
I'm no PGP expert, hopefully we can trust the GH instructions are giving us good advice so probably safest just to do what they say.

After adding the public key, GH will make it available at https://github.com/plantuml.gpg so anyone wanting to verify a jar can get the key from a trusted source.

**Reply by arnaudroques** [reply]
So I've imported a public key which [is available here](https://github.com/arnaudroques.gpg).
I am not sure I can make it available at https://github.com/plantuml.gpg because "plantuml" is an Organization account.

**Reply by matthew16550** [reply]
Oops sorry my bad, I do not see how to add keys to an org account either.  Guess their main purpose in GH is for individual users signing commits.

I was thinking https://github.com/plantuml.gpg would be a trustable location for anyone verifying a jar. Instead, we could put the public key on some key servers as recommended in the [central docs](https://central.sonatype.org/publish/requirements/gpg/#distributing-your-public-key).  And somewhere on https://plantuml.com/ would be nice.  But that could all wait until we finish updating the workflow.

To sign the jars, the workflow will need access to a private key and passphrase via two more secrets.

It is best if the jar signing key is different from the key on your `arnaudroques` account because they are separate things.  One is workflows making verifiable jars, the other is you making verifiable commits.

The key could have a User ID something like `PlantUML JAR Signing Key <plantuml@gmail.com>`

**Reply by arnaudroques** [reply]
> Instead, we could put the public key on some key servers as recommended in the [central docs](https://central.sonatype.org/publish/requirements/gpg/#distributing-your-public-key). And somewhere on https://plantuml.com/ would be nice. But that could all wait until we finish updating the workflow.

Sure, let's go step by step.



> It is best if the jar signing key is different from the key on your `arnaudroques` account because they are separate things. One is workflows making verifiable jars, the other is you making verifiable commits.
> 
> The key could have a User ID something like `PlantUML JAR Signing Key <plantuml@gmail.com>`

Understood. These is two separate things. But as you said, let's finish the workflow first.

**Comment by matthew16550** [comment]
`oss-parent` is [deprecated](https://central.sonatype.org/publish/publish-maven/#deprecated-oss-parent), it is not doing much so we can move the config into our own `pom.xml`.

One aspect of `oss-parent` is preventing use of some old problematic Maven versions.  I think we can do better by using [Maven Wrapper](https://github.com/takari/maven-wrapper#legacy-documentation) to ensure we are all on the same version.  It means we do `./mvnw` instead of `mvn` and IntelliJ / Eclipse might need an option changed to use the wrapper.

@arnaudroques  ?

**Reply by arnaudroques** [reply]
Well, I'm ok to remove `oss-parent`. However, it looks like [Maven Wrapper](https://github.com/takari/maven-wrapper#legacy-documentation) is also frozen, so I am uncomfortable about using it. And since a lot a project are using PlantUML library, is there an impact on all those projects ?

So ok to move the config into our own `pom.xml`. If it's too complex, we could also forget about old Maven versions and impose a new version of Maven to build this project.

Another idea: let's keep all simple: can we remove javadoc generation and jar signing? Would that help?

@hallvard Any idea ?

**Reply by hallvard** [reply]
The maven setup is partly for internal use, including signing and releasing, and partly to allow others to build their own copy, where they don't need to sign and release. Can signing and releasing be separated out?

I've never used the wrapper, but I'm all for requiring a pretty new version of maven.

**Reply by soloturn** [reply]
maven wrapper is now at apache: https://github.com/apache/maven-wrapper. maven version upgrades are rather rare, so the benefit of using it might be minor.

**Comment by matthew16550** [comment]
@arnaudroques sorry `maven-release-plugin` just accidentally pushed direct to the `plantuml` repo.  Forgot I had permissions there!

It made a spammy branch and a `v1.2021.15` tag which I've both deleted.  It didn't make a release.

**Comment by arnaudroques** [comment]
"With great power comes great responsibility" :-)


> It made a spammy branch and a `v1.2021.15` tag which I've both deleted. It didn't make a release.

No worries. Thanks for cleaning up!

**Comment by matthew16550** [comment]
Been down a few rabbit holes with this, it is a confusing jigsaw puzzle.

The whole thing seems to be working now up to the point of deploying to Central (which I have not tried).  But the PR would be very confusing and I hope to find some simplifications by breaking this into smaller PRs.

I _think_ the simplest first step is to set up jar signing in GH and publish the signature files in the `snapshot` releases.

We could also publish them in the main releases on GH but at this step they would have a different signing key from the signatures on Central so that might cause confusion.

@arnaudroques ?

**Reply by matthew16550** [reply]
@arnaudroques ⬆️ ? 
😄

**Reply by arnaudroques** [reply]
Sorry for the delay...



> I _think_ the simplest first step is to set up jar signing in GH and publish the signature files in the `snapshot` releases.

I agree, let's start by signing in GH.



> We could also publish them in the main releases on GH but at this step they would have a different signing key from the signatures on Central so that might cause confusion.

This will be a second step. Using a different signing key on Central will be ok.

**Comment by arnaudroques** [comment]
So there is now a secret ARTIFACT_SIGNING_KEY which looks like:

```
-----BEGIN PGP PRIVATE KEY BLOCK-----

XXXXX
-----END PGP PRIVATE KEY BLOCK-----
```

and another secret ARTIFACT_SIGNING_PASSPHRASE which looks like `this is a secret passphrase`.

The key used id is `PlantUML JAR Signing Key <plantuml@gmail.com>`

The public key is available at https://plantuml.com/pgp_plantuml_jar_signing_key.txt

**Reply by matthew16550** [reply]
I manually triggered the CI workflow, it looks to be working fine.  Signature files are in the snapshot and verify ok.

**Reply by arnaudroques** [reply]
Great! Maybe we should put the [public key](https://plantuml.com/pgp_plantuml_jar_signing_key.txt) in the [repository itself](https://github.com/plantuml/plantuml).
Does it make sense ?

**Reply by matthew16550** [reply]
Yes I guess that makes sense.  It is certainly not a secret.  I'm not an expert in this stuff, just following my nose.  But I don't see a downside and it is clearly a trustable location.

**Comment by matthew16550** [comment]
Hi @arnaudroques, just letting you know I haven't forgotten the rest of this!  Hopefully will have time for it in the next few weeks.

**Reply by arnaudroques** [reply]
Many thanks Matthew for the update!
Take your time; All this is a long-distance race but must remain a pleasure :-)

**Comment by arnaudroques** [comment]
There is another ongoing try here https://github.com/plantuml/plantuml/pull/1135.
We have some issue on publishing to OSSRH so any help is wecome :-)
Thanks,
